core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -2,105 +2,232 @@
 
 //! This module contains permissions related Iroha functionality.
 
-use std::{fmt::Debug, marker::PhantomData, sync::Arc};
+use std::{fmt::Debug, marker::PhantomData};
 
 pub use checks::*;
-pub use combinators::ValidatorApplyOr as _;
-use error::*;
 pub use has_token::*;
 use iroha_data_model::prelude::*;
-use iroha_macro::FromVariant;
 use iroha_schema::IntoSchema;
-pub use is_allowed::*;
 use parity_scale_codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
 
 use crate::wsv::WorldStateView;
 
-pub mod builder;
 mod checks;
 pub mod combinators;
 mod has_token;
-mod is_allowed;
+pub mod judge;
 pub mod roles;
 
-/// Result type for permission validators
+/// Result type associated with permission validators
 pub type Result<T> = std::result::Result<T, DenialReason>;
 
-/// Operation for which the permission should be checked.
-pub trait NeedsPermission: Debug {}
+/// Operation for which the permission should be checked
+pub trait NeedsPermission: Debug {
+    /// Get the type of validator required to check the operation
+    ///
+    /// Accepts `self` because of the [`NeedsPermissionBox`]
+    fn required_validator_type(&self) -> ValidatorType;
+}
 
-impl NeedsPermission for Instruction {}
+impl NeedsPermission for Instruction {
+    fn required_validator_type(&self) -> ValidatorType {
+        ValidatorType::Instruction
```

Comment:
Could this be a constant instead?

Reply:
Unfortunately functions can not be const in traits and traits impl

Reply:
I meant 
```rust 
const REQUIRED_VALIDATOR_TYPE: ValidatorType = ValidatorType::Instruction
``` 

and replace each `self.required_validator_type()` with `Self::REQUIRED_VALIDATOR_TYPE`.

Reply:
Hmm, but `NeedsPermissionBox` is useless right now... But may be usefull for deseriasation.
If I won't find `NeedsPermissionBox` useful while implementing #2199 I'll remove it and make this be const as you proposed